sandbox/sandbox/src/embedded_executor.rs

```diff
@@ -18,325 +18,407 @@
+            instance: self.instance,
+            _marker: PhantomData,
+        }
+    }
 }
 
-impl<T> super::SandboxInstance<T> for Instance<T> {
+impl<State: 'static> super::SandboxInstance<State> for Instance<State> {
     type Memory = Memory;
-    type EnvironmentBuilder = EnvironmentDefinitionBuilder<T>;
+    type EnvironmentBuilder = EnvironmentDefinitionBuilder<State>;
 
     fn new(
+        mut store: &mut Store<State>,
         code: &[u8],
-        env_def_builder: &EnvironmentDefinitionBuilder<T>,
-        state: &mut T,
-    ) -> Result<Instance<T>, Error> {
-        let module = Module::from_buffer(code).map_err(|_| Error::Module)?;
-        let not_started_instance =
-            ModuleInstance::new(&module, env_def_builder).map_err(|_| Error::Module)?;
-
-        let defined_host_functions = env_def_builder.defined_host_functions.clone();
-        let instance = {
-            let mut externals = GuestExternals {
-                state,
-                defined_host_functions: &defined_host_functions,
-            };
-            not_started_instance
-                .run_start(&mut externals)
-                .map_err(|_| Error::Execution)?
-        };
+        env_def_builder: &Self::EnvironmentBuilder,
+    ) -> Result<Instance<State>, Error> {
+        let module = Module::new(store.engine(), code).map_err(|e| {
+            log::error!(target: TARGET, "Failed to create module: {e}");
+            Error::Module
+        })?;
+        let mut linker = Linker::new(store.engine());
+
+        for import in module.imports() {
+            let module = import.module().to_string();
+            let name = import.name().to_string();
+            let key = (module.clone(), name.clone());
+
+            match import.ty() {
+                ExternType::Global(_) | ExternType::Table(_) => {}
+                ExternType::Memory(_mem_ty) => {
+                    let mem = env_def_builder
+                        .map
+                        .get(&key)
+                        .cloned()
+                        .and_then(|val| val.memory())
+                        .ok_or(Error::Module)?
+                        .memref;
+
+                    let mem = wasmi::Extern::Memory(mem);
+                    linker
+                        .define(&module, &name, mem)
+                        .map_err(|_| Error::Module)?;
+                }
+                ExternType::Func(func_ty) => {
+                    let func_ptr = env_def_builder
+                        .map
+                        .get(&key)
+                        .cloned()
+                        .and_then(|val| val.host_func())
+                        .ok_or(Error::Module)?;
+
+                    let func = wasmi::Func::new(
+                        &mut store,
+                        func_ty.clone(),
+                        move |caller, params, results| {
+                            let gas = caller
+                                .get_export(GLOBAL_NAME_GAS)
```

Comment:
so for now we have unified approach on globals for both backends, correct?

Reply:
We do